### PR TITLE
Add --random-fully to MASQ iptables rules to mitigate conntrack issues

### PR DESCRIPTION
**Pull request by coufalja**
To mitigate conntrack issues in Kubernetes.
See: https://tech.xing.com/a-reason-for-unexplained-connection-timeouts-on-kubernetes-docker-abd041cf7e02 and kubernetes/kubernetes#56903 for details.

### CODE REVIEW

**Comment by coufalja** [comment]
Docker image for testing purposes: `wanderadock/kube-router:amd64-wnd-randomize`

**Comment by murali-reddy** [comment]
@coufalja thanks for the PR. `--random-fully` option make sense. Changes look good but on upgrade we would end up with two rules. Old rule plus new rule with `--random-fully` option set. Are you expecting users to flush the iptables chains?

File: Gopkg.lock [hunk 1]
```diff
@@ -98,12 +98,12 @@
...
 [[projects]]
-  digest = "1:df4af3a8e15cd72b068b8d898598001935edce4eac4eb5a3d7fd906036123ff4"
+  digest = "1:2debc00eaf8ffdbfb0570a791a84e64a3cfc146f5e61b54d4abd65c2de364552"
   name = "github.com/coreos/go-iptables"
   packages = ["iptables"]
   pruneopts = "UT"
-  revision = "259c8e6a4275d497442c721fa52204d7a58bde8b"
-  version = "v0.2.0"
+  revision = "f901d6c2a4f2a4df092b98c33366dfba1f93d7a0"
+  version = "v0.4.5"
```
Comment (mrueg) [inline Gopkg.lock:106]: you're setting a constraint to v0.4.1 but this updates it to 0.4.5?

**Comment by coufalja** [comment]
@murali-reddy The way I deployed was going through full rolling-update of a cluster so I didn't hit the issue. But yeah It would be great to actually allow users to update in place. What needs to be done? Adding the old one to pkg/controllers/routing/pod_egress.go:85 and pkg/controllers/proxy/network_services_controller.go:1268 ?

File: Gopkg.lock (hunk 1 above)
Comment (coufalja) [reply Gopkg.lock:106]: Huh? I just run dep I have no idea how it happened, will fix.

Comment (coufalja) [reply Gopkg.lock:106]: Oh I see... actually the constraint in dep reads like this `^0.4.1` so 0.4.5 is right

File: pkg/controllers/proxy/network_services_controller.go [hunk 2]
```diff
@@ -1269,6 +1276,11 @@ func (nsc *NetworkServicesController) deleteBadMasqueradeIptablesRules() error {
 		{"-m", "ipvs", "--ipvs", "--vdir", "ORIGINAL", "--vmethod", "MASQ", "-m", "comment", "--comment", "", "!", "-s", nsc.podCidr, "!", "-d", nsc.podCidr, "-j", "MASQUERADE"},
 	}
 
+	// If random fully is supported remove the original rule as well
+	if iptablesCmdHandler.HasRandomFully() {
+		argsBad = append(argsBad, []string{"-m", "ipvs", "--ipvs", "--vdir", "ORIGINAL", "--vmethod", "MASQ", "-m", "comment", "--comment", "", "-j", "SNAT", "--to-source", nsc.nodeIP.String()})
```
Comment (aauren) [inline pkg/controllers/proxy/network_services_controller.go:1281]: Don't you have to cover the podCidr case here too? (i.e. the 2nd change to this file above?)

Comment (coufalja) [reply pkg/controllers/proxy/network_services_controller.go:1281]: Thanks you are absolutely right!

Comment (coufalja) [reply pkg/controllers/proxy/network_services_controller.go:1281]: Done

File: pkg/controllers/proxy/network_services_controller.go [hunk 3]
```diff
@@ -1269,6 +1276,16 @@ func (nsc *NetworkServicesController) deleteBadMasqueradeIptablesRules() error {
 		{"-m", "ipvs", "--ipvs", "--vdir", "ORIGINAL", "--vmethod", "MASQ", "-m", "comment", "--comment", "", "!", "-s", nsc.podCidr, "!", "-d", nsc.podCidr, "-j", "MASQUERADE"},
 	}
 
+	// If random fully is supported remove the original rules as well
+	if iptablesCmdHandler.HasRandomFully() {
+		argsBad = append(argsBad, []string{"-m", "ipvs", "--ipvs", "--vdir", "ORIGINAL", "--vmethod", "MASQ", "-m", "comment", "--comment", "", "-j", "SNAT", "--to-source", nsc.nodeIP.String()})
+
```
Comment (aauren) [inline pkg/controllers/proxy/network_services_controller.go:1282]: It looks like there is some dangling whitespace on this line that gofmt is screaming about. Once that is fixed though, I don't see any other problems with this PR.

Comment (coufalja) [reply pkg/controllers/proxy/network_services_controller.go]: Oh missed the whitespace, done

**Comment by coufalja** [comment]
@aauren Thanks, just a question when I can expect this to be released? in the 1.1 or you plan to release 1.0.1 with a round of minor fixes?

**Comment by mrueg** [comment]
This would be an additional feature, so it will be in 1.1

**Comment by murali-reddy** [comment]
> I tested this locally on our nodes with both iptables that are able to support --random-full and iptables that are not. I also tested the upgrade procedure and found that it cleaned the nodes appropriately when upgrading from iptables that didn't support it to iptables that did support it.

thanks @aauren

**Pull request merged**